core/src/cluster_slot_state_verifier.rs

```diff
@@ -33,147 +45,185 @@ impl SlotStateUpdate {
-        // version.
-        // In both cases, mark the progress map for this slot as duplicate
-        return vec![ResultingStateChange::MarkSlotDuplicate(*bank_frozen_hash)];
+    assert!(cluster_duplicate_confirmed_hash.is_some());
+    if !is_dead && bank_frozen_hash == Hash::default() {
+        return vec![];
     }
 
-    vec![]
+    let mut state_changes = vec![];
+    check_duplicate_confirmed_hash_against_frozen_hash(
+        &mut state_changes,
+        slot,
+        cluster_duplicate_confirmed_hash,
+        bank_frozen_hash,
+        is_dead,
+        is_slot_duplicate,
+    );
+
+    state_changes
 }
 
-// Called when we receive either:
-// 1) A duplicate slot signal from WindowStage,
-// 2) Confirmation of a slot by observing votes from replay or gossip.
-//
-// This signals external information about this slot, which affects
-// this validator's understanding of the validity of this slot
-fn on_cluster_update(
-    slot: Slot,
-    bank_frozen_hash: &Hash,
-    cluster_duplicate_confirmed_hash: Option<&Hash>,
-    is_slot_duplicate: bool,
-    is_dead: bool,
-) -> Vec<ResultingStateChange> {
-    if is_dead {
-        on_dead_slot(
-            slot,
-            bank_frozen_hash,
-            cluster_duplicate_confirmed_hash,
-            is_slot_duplicate,
-            is_dead,
-        )
-    } else if *bank_frozen_hash != Hash::default() {
-        // This case is mutually exclusive with is_dead case above because if a slot is dead,
-        // it cannot have  been frozen, and thus cannot have a non-default bank hash.
-        on_frozen_slot(
-            slot,
-            bank_frozen_hash,
-            cluster_duplicate_confirmed_hash,
-            is_slot_duplicate,
-            is_dead,
-        )
+fn on_duplicate(args: SlotStateHandlerArgs) -> Vec<ResultingStateChange> {
+    let SlotStateHandlerArgs {
+        bank_frozen_hash,
+        cluster_duplicate_confirmed_hash,
+        is_dead,
+        ..
+    } = args;
+
+    if !is_dead && bank_frozen_hash == Hash::default() {
+        return vec![];
+    }
+
+    if let Some(cluster_duplicate_confirmed_hash) = cluster_duplicate_confirmed_hash {
+        // If the cluster duplicate_confirmed some version of this slot, then
+        // confirm our version agrees with the cluster,
+        if cluster_duplicate_confirmed_hash != bank_frozen_hash {
+            // If we detected a duplicate, but have not yet seen any version
+            // of the slot duplicate_confirmed (i.e. block above did not execute), then
+            // remove the slot from fork choice until we get confirmation.
+
+            // If we get here, we either detected duplicate from
+            // 1) WindowService
+            // 2) A gossip duplicate_confirmed version that didn't match our frozen
+            // version.
+            // In both cases, mark the progress map for this slot as duplicate
+            return vec![ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash)];
+        }
```

Comment:
might be cleaner like this since both cases result in marking duplicate
```suggestion
    if let Some(cluster_duplicate_confirmed_hash) = cluster_duplicate_confirmed_hash  && cluster_duplicate_confirmed_hash == bank_frozen_hash {
       return vec![]
    }
    vec![ResultingStateChange::MarkSlotDuplicate(bank_frozen__hash)];
}
```

Reply:
yeah a lot better, thanks